pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,201 @@
+package logger
+
+import (
+	"container/list"
+	"fmt"
+	"go.uber.org/zap/zapcore"
+	"log"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+)
+
+
+var UnixErrorLogPath = "/opt/aws/aws-otel-collector/logs/aws-otel-collector-error.log"
+
+var WindowsErrorLogPath = "C:\\ProgramData\\Amazon\\AWSOTelCollector\\Logs\\aws-otel-collector-error.log"
+
+var errorLogfile = getErrorLogFilePath()
+
+var errorLogMaxAge = 10
+
+var queue = list.New()
+var errorLogChannel = make(chan zapcore.Entry, 10)
+var ticker = time.NewTicker(10 * time.Minute)
+var quit = make(chan struct{})
+var kilobyte = 1024
+var errorFile *os.File
+
+// WriteError func could write the new error log message into file
+func WriteError(newErrorLog zapcore.Entry, errorLogfile string) (n int, err error) {
+	var byteNum int
+	if errorFile == nil {
+		if err = openExistingOrNewErrorFile(); err != nil {
+			return 0, err
+		}
+	}
+	err = errorFile.Truncate(0)
+	if err != nil {
+		return byteNum, err
+	}
+	// iterate errors in queue, if the old error's message same as the new error's message, skip it,
+	// otherwise write it into the file and queue
+	var backupQueue = list.New()
+	for queue.Len() > 0 {
+		e := queue.Front().Value.(zapcore.Entry)
+		if e.Message != newErrorLog.Message {
+			n, err = errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				return byteNum, err
+			}
+			byteNum += n
+			backupQueue.PushBack(e)
+		}
+		queue.Remove(queue.Front())
+	}
+	backupQueue.PushBack(newErrorLog)
+	queue = backupQueue
+	n, err = errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+		newErrorLog.Time, newErrorLog.Level, newErrorLog.Caller, newErrorLog.Message)))
+	if err != nil {
+		return byteNum, err
+	}
+	byteNum += n
```

Comment:
Wondering rather than doing a dequeue and enqueuing again, keeping a map and list wouldn't b e efficient. map for lookup and list for deleting the item in order?